Add ability to add custom links to User menu via server CLI or ENV vars

jaredscheib opened the pull request:
- [x] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] Rebased/mergable
  - [x] Tests pass
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Connect #1550 

### The problem
The side nav menu in the client only included pre-determined links, and didn't allow for any custom links inputted by the user.

### The Solution
Allow the user to add custom links to their User menu in the client side nav, via two methods:
  - CLI flag on server process run, e.g. `--custom-link=InfluxData:https://influxdata.com --custom-link=Pineapple:http://reddit.com/r/KnightsOfPineapple/`
  - ENV var in server environment, e.g. `export CUSTOM_LINKS=InfluxData:https://influxdata.com,Pineapple:http://reddit.com/r/KnightsOfPineapple/`

<img width="215" alt="screen shot 2017-06-26 at 11 47 40 am" src="https://user-images.githubusercontent.com/6403018/27555035-493b9998-5a65-11e7-97a7-ab73a137b986.png">

jaredscheib commented on ui/src/utils/ajax.js:
```diff
@@ -9,11 +9,15 @@ const addBasepath = (url, excludeBasepath) => {
-const generateResponseWithLinks = (response, {auth, logout, external}) => ({
+const generateResponseWithLinks = (
+  response,
+  {auth, logout, external, custom}
```
custom looks obsolete. should be subsumed by external.

goller commented on server/links.go:
```diff
@@ -0,0 +1,41 @@
+		}
+		_, err := url.Parse(link)
+		if err != nil {
+			return customLinks, err
```
Minor detail, but would you return `nil` rather than customLinks

jaredscheib replied:
Ah, interesting -- yes. Shall I also return `nil` instead of `customLinks` in the two conditionals just above that?

goller replied:
Yuppers

jaredscheib replied:
Cool - thanks

timraymond commented on server/links.go:
```diff
@@ -0,0 +1,41 @@
+// CustomLink is a handler that returns a custom link to be used in server's routes response, within ExternalLinks
+type CustomLink struct {
+	Name string `json:"name,omitempty"`
+	URL  string `json:"url,omitempty"`
```
Is there a useful situation where a custom link is present but with only a `Name` or `URL`? If not, we can probably remove `omitempty` from both of these and tighten up the guarantees that we provide to clients.

jaredscheib replied:
Cool -- good thinking. There isn't, and it's currently written to throw an error in that circumstances, so there should not be an `omitempty`.

cryptoquick commented on ui/src/side_nav/containers/SideNav.js:
```diff
@@ -25,6 +25,37 @@ const SideNav = React.createClass({
+  },
+
+  renderUserMenuBlockWithCustomLinks(customLinks, logoutLink) {
+    return [<NavHeader key={0} title="User" />].concat(
```
Have you looked into using the array spread operator to simplify these concats?

i.e., 

```javascript
[<NavHeader key={0} title="User" />, ...customLinks.map(), <NavListItem>Logout</NavListItem>]
```

jaredscheib replied:
Ah cool, will change -- thanks

jaredscheib commented:
Also, two questions @goller:
- I'd like to add two custom links to routes_test.go, but this test will fail sometimes, presumably due to the map that's being passed into NewCustomLink having a random order of keys, and thus its output's order varies. Would it be reasonable to refactor NewCustomLinks to ensure consistent order?
- For `TestNewCustomLinks`, it feels weird to me to test a negative error condition -- does that seem right to you, or is there a better way to do this?

jaredscheib commented on server/links.go:
```diff
@@ -0,0 +1,41 @@
+// NewCustomLinks transforms `--custom-link` CLI flag data or `CUSTOM_LINKS` ENV
+// var data into a data structure that the Chronograf client will expect
+func NewCustomLinks(links map[string]string) ([]CustomLink, error) {
+	var customLinks []CustomLink
```
Good call -- thanks

jaredscheib commented:
Thanks @timraymond. I decided to keep the `routes_test.go` as-is, since ensuring order and changing the test to match would a larger undertaking. Per an offline suggestion by @goller, I went ahead and sorted by user nav items alphabetically on the client-side, since the UI is where the UX for alphabetical would be relevant.

Addressed all comments from reviews in the commits that followed.

nhaugo commented:
@jaredscheib using the option it occurs to me that ":" is maybe not the best separator. The option `https://google.com" shows up as a menu item "https" with a link to `//google.com`.

At a minimum the help message "Custom link to be added to the client User menu [$CUSTOM_LINKS]" should probably explain the syntax and have an example, e.g.
"Custom link with a name to be added to the client User menu, the name and link are separated by a :, e.g. influxdata:https://influxdata.com [$CUSTOM_LINKS]"
